### PR TITLE
Feature/time interpolation

### PR DESCRIPTION
**Pull request by emiax**
- Ability to interpolate in time
- Buffered rendering of planet projections
- Works with parallel connection

### CODE REVIEW

File: data/assets/util/default_keybindings.asset [hunk 1]
```diff
@@ -52,9 +52,9 @@ openspace.setPropertyValueSingle('Modules.ImGUI.Main.IsHidden', b);]],
     },
     {
         Key = "SPACE",
-        Command = "openspace.time.togglePause()",
+        Command = "openspace.time.interpolateTogglePause()",
```
Comment (alexanderbock) [inline data/assets/util/default_keybindings.asset]: Adding Shift+Space for immediate pause as well?

File: include/openspace/interaction/keyframenavigator.h [hunk 2]
```diff
@@ -48,7 +48,7 @@ class KeyframeNavigator {
     Timeline<CameraPose>& timeline();
 
     void addKeyframe(double timestamp, KeyframeNavigator::CameraPose pose);
-    void removeKeyframesAfter(double timestamp);
+    void removeKeyframesAfter(double timestamp, bool inclusive = false);
```
Comment (alexanderbock) [inline include/openspace/interaction/keyframenavigator.h]: Prefer:
```
#include <ghoul/misc/boolean.h>
BooleanType(Inclusive);
void removeKeyframesAfter(double timestamp, Inclusive inclusive = Inclusive::No);
```
as it makes the caller site easier to understand

File: include/openspace/network/messagestructures.h [hunk 3]
```diff
@@ -229,11 +226,11 @@ struct ScriptMessage {
 
     std::string _script;
 
-    void serialize(std::vector<char> &buffer){
+    void serialize(std::vector<char> &buffer) {
```
Comment (alexanderbock) [inline include/openspace/network/messagestructures.h]: Can be `const` as well?

File: include/openspace/network/messagestructures.h [hunk 4]
```diff
@@ -229,11 +226,11 @@ struct ScriptMessage {
...
         buffer.insert(buffer.end(), _script.begin(), _script.end());
     };
 
-    void deserialize(const std::vector<char> &buffer){
+    void deserialize(const std::vector<char> &buffer) {
```
Comment (alexanderbock) [inline include/openspace/network/messagestructures.h]: Can be `const` as well?

File: include/openspace/network/parallelconnection.h [hunk 5]
```diff
@@ -83,6 +84,7 @@ class ParallelConnection  {
 
     ParallelConnection::Message receiveMessage();
 
+    static unsigned int protocolVersion();
```
Comment (alexanderbock) [inline include/openspace/network/parallelconnection.h]: How about making the anonymous namespace variable a class variable instead and not have the indirection through this function?

File: include/openspace/util/timemanager.h [hunk 6]
```diff
@@ -32,41 +32,107 @@
...
     void removeTimeChangeCallback(CallbackHandle handle);
     void removeDeltaTimeChangeCallback(CallbackHandle handle);
-
+    void removeTimeJumpCallback(CallbackHandle handle);
+    void removeTimelineChangeCallback(CallbackHandle handle);
 private:
+    void progressTime(double dt);
+    void applyKeyframeData(const TimeKeyframeData& keyframe);
+    TimeKeyframeData interpolate(
+        const Keyframe<TimeKeyframeData>& past,
```
Comment (alexanderbock) [inline include/openspace/util/timemanager.h]: For style:  This line can be in the same line as the function definition;  the `time` can be moved in with the `future` line, too

File: modules/spacecraftinstruments/rendering/renderablefov.cpp [hunk 7]
```diff
@@ -892,16 +892,18 @@ void RenderableFov::render(const RenderData& data, RendererTasks&) {
 void RenderableFov::update(const UpdateData& data) {
     _drawFOV = false;
     if (openspace::ImageSequencer::ref().isReady()) {
-        _drawFOV = ImageSequencer::ref().isInstrumentActive(_instrument.name);
+        _drawFOV = ImageSequencer::ref().isInstrumentActive(data.time.j2000Seconds(),
```
Comment (alexanderbock) [inline modules/spacecraftinstruments/rendering/renderablefov.cpp]:
```
_drawFOV = ImageSequencer::ref().isInstrumentActive(
    data.time.j2000Seconds(),
    _instrument.name
);
```

File: modules/spacecraftinstruments/rendering/renderableplaneprojection.cpp [hunk 8]
```diff
@@ -105,7 +105,8 @@ void RenderablePlaneProjection::deinitializeGL() {
 }
 
 void RenderablePlaneProjection::render(const RenderData& data, RendererTasks&) {
-    bool active = ImageSequencer::ref().isInstrumentActive(_instrument);
+    bool active = ImageSequencer::ref().isInstrumentActive(data.time.j2000Seconds(),
```
Comment (alexanderbock) [inline modules/spacecraftinstruments/rendering/renderableplaneprojection.cpp]: Same as above

File: modules/spacecraftinstruments/rendering/renderableplanetprojection.cpp [hunk 9]
```diff
@@ -107,6 +107,26 @@ namespace {
         "Ambient Brightness",
         "This value determines the ambient brightness of the dark side of the planet."
     };
+
+    const openspace::properties::Property::PropertyInfo MaxProjectionsPerFrameInfo = {
```
Comment (alexanderbock) [inline modules/spacecraftinstruments/rendering/renderableplanetprojection.cpp]: `constexpr`

File: modules/spacecraftinstruments/rendering/renderableplanetprojection.cpp [hunk 10]
```diff
@@ -559,23 +606,31 @@ void RenderablePlanetProjection::render(const RenderData& data, RendererTasks&)
...
         for (const Image& img : _imageTimes) {
+            if (nPerformedProjections >= _maxProjectionsPerFrame) {
+                break;
+            }
             RenderablePlanetProjection::attitudeParameters(img.timeRange.start);
             imageProjectGPU(_projectionComponent.loadProjectionTexture(img.path));
+            ++nPerformedProjections;
         }
-        _shouldCapture = false;
+        _imageTimes.erase(_imageTimes.begin(),
```
Comment (alexanderbock) [inline modules/spacecraftinstruments/rendering/renderableplanetprojection.cpp]:
```
_imageTimes.erase(
   _imageTimes.begin(),
   _imageTimes.begin() + nPerformedProejctions
);
```

File: modules/spacecraftinstruments/rendering/renderableplanetprojection.cpp [hunk 11]
```diff
@@ -720,26 +775,59 @@ void RenderablePlanetProjection::update(const UpdateData& data) {
...
+    if (it != _imageTimes.end()) {
+        _imageTimes.erase(it, _imageTimes.end());
+    }
+}
+
+void RenderablePlanetProjection::insertImageProjections(const std::vector<Image>& images)
+{
+    _imageTimes.insert(_imageTimes.end(),
+        images.begin(),
+        images.end());
```
Comment (alexanderbock) [inline modules/spacecraftinstruments/rendering/renderableplanetprojection.cpp]: `);` on new line

File: src/scene/scene.cpp [hunk 12]
```diff
@@ -22,23 +22,31 @@
  * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                                         *
  ****************************************************************************************/
 
+#pragma optimize ("", off)
```
Comment (alexanderbock) [inline src/scene/scene.cpp]: Remove? optimize an fstream include

File: src/util/timemanager.cpp [hunk 13]
```diff
@@ -29,22 +29,84 @@
...
 #include <openspace/util/timeline.h>
 
+#include <ghoul/logging/logmanager.h>
+
+namespace {
+    // Properties for time interpolation
+    // These are used when setting the time from lua time interpolation functions,
+    // when called without arguments.
+
+    static const openspace::properties::Property::PropertyInfo
```
Comment (alexanderbock) [inline src/util/timemanager.cpp]: `static const => constexpr`

File: src/util/timemanager.cpp [hunk 14]
```diff
@@ -29,22 +29,84 @@
...
+        DefaultUnpauseInterpolationDurationInfo = {
+        "DefaultUnpauseInterpolationDuration",
+        "Default Unpause Interpolation Duration",
+        "The default duration taken to transition to the unpaused state, "
+        "when interpolating"
+    };
+
+    const char* _loggerCat = "TimeManager";
```
Comment (alexanderbock) [inline src/util/timemanager.cpp]: `constexpr const char*`

File: src/util/timemanager.cpp [hunk 15]
```diff
@@ -59,133 +121,264 @@ void TimeManager::preSynchronization(double dt) {
...
+}
 
-        // Do not interpolate with time jumping keyframes.
-        // Instead, wait until their timestamp and apply them directly.
-        if (nextTime.timeJumped()) {
-            time().advanceTime(dt);
-            return;
-        }
+void TimeManager::progressTime(double dt) {
+    static_cast<Time&>(_integrateFromTime) = _currentTime;
```
Comment (alexanderbock) [inline src/util/timemanager.cpp]: That looks really ugly;  is there a better way of expressing it without doing a cast on the lvalue?

File: src/util/timemanager.cpp [hunk 16]
```diff
@@ -59,133 +121,264 @@ void TimeManager::preSynchronization(double dt) {
...
+        keyframes.end();
+
+    const bool hasConsumedLastPastKeyframe = hasPastKeyframes ?
+        (lastPastKeyframe->timestamp <= _latestConsumedTimestamp) :
+        true;
+
+    if (hasFutureKeyframes && hasPastKeyframes && !firstFutureKeyframe->data.jump) {
+        // If keyframes exist before and after this frame,
+        // interpolate between those.
+        TimeKeyframeData interpolated =
```
Comment (alexanderbock) [inline src/util/timemanager.cpp]: Prefer:
```
TimeKeyframeData interpolated = interpolate(
    *lastPastKeyframe,
    *firstFutureKeyframe,
    now
);
```

File: src/util/timemanager.cpp [hunk 17]
```diff
@@ -59,133 +121,264 @@ void TimeManager::preSynchronization(double dt) {
...
+            interpolate(*lastPastKeyframe, *firstFutureKeyframe, now);
+        static_cast<Time&>(_currentTime).setTime(interpolated.time.j2000Seconds());
+        _deltaTime = interpolated.delta;
+    } else if (!hasConsumedLastPastKeyframe) {
+        applyKeyframeData(lastPastKeyframe->data);
+    } else if (!isPaused()) {
+        // If there are no keyframes to consider
+        // and time is not paused, just advance time.
+        _deltaTime = _targetDeltaTime;
+        static_cast<Time&>(_currentTime).advanceTime(dt * _deltaTime);
```
Comment (alexanderbock) [inline src/util/timemanager.cpp]: The multiple static_cast<Time&> are confusing.  Is there a better way to get the data out of the `SyncData` class?

File: src/util/timemanager.cpp [hunk 18]
```diff
@@ -59,133 +121,264 @@ void TimeManager::preSynchronization(double dt) {
...
-        }
+    if (hasPastKeyframes) {
+        _latestConsumedTimestamp = lastPastKeyframe->timestamp;
+    }
+}
 
-        const double t0 = now - dt;
-        const double t1 = now;
-        const double t2 = next.timestamp;
+TimeKeyframeData TimeManager::interpolate(
```
Comment (alexanderbock) [inline src/util/timemanager.cpp]:
```
TimeKeyframeData TimeManager::interpolate(const Keyframe<TimeKeyframeData>& past,
                                          const Keyframe<TimeKeyframeData>& future,
                                          double appTime)
{
```

File: src/scene/scene.cpp (hunk 12 above)
Comment (emiax) [reply src/scene/scene.cpp]: No clue how that made it in there...

**Pull request merged**